File: XCTestBootstrapTests/Tests/FBTestConfigurationTests.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <objc/runtime.h>

#import <XCTestBootstrap/XCTestBootstrap.h>
#import <XCTestPrivate/XCTestConfiguration.h>
#import <XCTestPrivate/XCTCapabilities.h>

@interface FBTestConfigurationTests : XCTestCase

@end

@implementation FBTestConfigurationTests

- (void)testSimpleConstructor
{
  XCTestConfiguration * xcTestConfig = [objc_lookUpClass("XCTestConfiguration") new];
  NSUUID *sessionIdentifier = [[NSUUID alloc] initWithUUIDString:@"E621E1F8-C36C-495A-93FC-0C247A3E6E5F"];
  FBTestConfiguration *testConfiguration = [FBTestConfiguration
    configurationWithSessionIdentifier:sessionIdentifier
    moduleName:@"Franek"
    testBundlePath:@"BundlePath"
    path:@"ConfigPath"
    uiTesting:YES
    xcTestConfiguration:xcTestConfig];

  XCTAssertTrue([testConfiguration isKindOfClass:FBTestConfiguration.class]);
  XCTAssertEqual(testConfiguration.sessionIdentifier, sessionIdentifier);
  XCTAssertTrue([testConfiguration isKindOfClass:FBTestConfiguration.class]);
  XCTAssertEqual(testConfiguration.testBundlePath, @"BundlePath");
  XCTAssertEqual(testConfiguration.path, @"ConfigPath");
  XCTAssertTrue(testConfiguration.shouldInitializeForUITesting);
  XCTAssertEqual(testConfiguration.xcTestConfiguration, xcTestConfig);
    
}

- (void)testSaveAs
{
  NSError *error;
  NSUUID *sessionIdentifier = NSUUID.UUID;
  NSString *someRandomPath = NSTemporaryDirectory();

  FBTestConfiguration *testConfiguration = [FBTestConfiguration
    configurationByWritingToFileWithSessionIdentifier:sessionIdentifier
    moduleName:@"ModuleName"
    testBundlePath:someRandomPath
    uiTesting:YES
    testsToRun:[NSSet set]
    testsToSkip:[NSSet set]
    targetApplicationPath:@"targetAppPath"
    targetApplicationBundleID:@"targetBundleID"
    testApplicationDependencies: nil
    automationFrameworkPath:nil
    reportActivities:NO
    error:&error];

  XCTAssertNil(error);
  XCTAssertNotNil(testConfiguration);
  XCTAssertTrue([NSFileManager.defaultManager fileExistsAtPath:testConfiguration.path]);

  XCTestConfiguration *xcTestConfig = testConfiguration.xcTestConfiguration;

  XCTAssertNotNil(xcTestConfig);
  XCTAssertEqual(xcTestConfig.productModuleName, @"ModuleName");
  XCTAssertEqualObjects(xcTestConfig.testBundleURL, [NSURL fileURLWithPath:someRandomPath]);
  XCTAssertEqual(xcTestConfig.initializeForUITesting, YES);
  XCTAssertEqual(xcTestConfig.targetApplicationPath, @"targetAppPath");
  XCTAssertEqual(xcTestConfig.targetApplicationBundleID, @"targetBundleID");
  XCTAssertEqual(xcTestConfig.reportActivities, NO);
  XCTAssertEqual(xcTestConfig.reportResultsToIDE, YES);
  
  NSDictionary *capabilities = @{@"XCTIssue capability": @1, @"ubiquitous test identifiers": @1};
  XCTAssertEqualObjects(xcTestConfig.IDECapabilities.capabilitiesDictionary, capabilities);
}

@end

File: XCTestBootstrapTests/Tests/FBTestConfigurationTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import XCTest
import XCTestBootstrap

final class FBTestConfigurationTests: XCTestCase {

  func testSimpleConstructor() {
    let xcTestConfig = FBTestConfigurationTestHelper.createXCTestConfiguration()
    let sessionIdentifier = UUID(uuidString: "E621E1F8-C36C-495A-93FC-0C247A3E6E5F")!
    let testConfiguration = FBTestConfigurationTestHelper.createTestConfiguration(
      withSessionIdentifier: sessionIdentifier,
      moduleName: "Franek",
      testBundlePath: "BundlePath",
      path: "ConfigPath",
      uiTesting: true,
      xcTestConfiguration: xcTestConfig
    )

    XCTAssertTrue(testConfiguration is FBTestConfiguration)
    XCTAssertEqual(testConfiguration.sessionIdentifier, sessionIdentifier)
    XCTAssertTrue(testConfiguration is FBTestConfiguration)
    XCTAssertEqual(testConfiguration.testBundlePath, "BundlePath")
    XCTAssertEqual(testConfiguration.path, "ConfigPath")
    XCTAssertTrue(testConfiguration.shouldInitializeForUITesting)
    XCTAssertTrue(testConfiguration.xcTestConfiguration === (xcTestConfig as AnyObject))
  }

  func testSaveAs() throws {
    let sessionIdentifier = UUID()
    let someRandomPath = NSTemporaryDirectory()

    let testConfiguration = try FBTestConfigurationTestHelper.createTestConfigurationByWritingToFile(
      withSessionIdentifier: sessionIdentifier,
      moduleName: "ModuleName",
      testBundlePath: someRandomPath,
      uiTesting: true,
      testsToRun: Set(),
      testsToSkip: Set(),
      targetApplicationPath: "targetAppPath",
      targetApplicationBundleID: "targetBundleID",
      testApplicationDependencies: nil,
      automationFrameworkPath: nil,
      reportActivities: false
    )

    XCTAssertTrue(FileManager.default.fileExists(atPath: testConfiguration.path!))

    let xcTestConfig = testConfiguration.xcTestConfiguration

    XCTAssertNotNil(xcTestConfig)
    XCTAssertEqual(FBTestConfigurationTestHelper.productModuleName(xcTestConfig), "ModuleName")
    XCTAssertEqual(FBTestConfigurationTestHelper.testBundleURL(xcTestConfig), URL(fileURLWithPath: someRandomPath))
    XCTAssertEqual(FBTestConfigurationTestHelper.initialize(forUITesting: xcTestConfig), true)
    XCTAssertEqual(FBTestConfigurationTestHelper.targetApplicationPath(xcTestConfig), "targetAppPath")
    XCTAssertEqual(FBTestConfigurationTestHelper.targetApplicationBundleID(xcTestConfig), "targetBundleID")
    XCTAssertEqual(FBTestConfigurationTestHelper.reportActivities(xcTestConfig), false)
    XCTAssertEqual(FBTestConfigurationTestHelper.reportResults(toIDE: xcTestConfig), true)

    let capabilities: NSDictionary = ["XCTIssue capability": 1, "ubiquitous test identifiers": 1]
    XCTAssertEqual(FBTestConfigurationTestHelper.ideCapabilitiesDictionary(xcTestConfig)! as NSDictionary, capabilities)
  }
}
